File: workflow/scripts/plot_alignment_rate.py
import sys
import os
import pandas as pd
import numpy as np
import general_functions as utils

# write stdout/stderr to log file
sys.stderr = open(snakemake.log[0], "w")

# create df to store alignment rates
df = pd.DataFrame(columns=["sample","alignment_rate"], index=np.arange(len(snakemake.input)))
samples = []
rates = []

for i in sorted(snakemake.input):
    # get sample name from file name
    sample = os.path.basename(i).replace(".log","")
    samples.append(sample)
    
    # extract alignment rate from file
    with open(i) as f:
        lines = f.readlines()
    rate = float([x for x in lines if "overall alignment rate" in x][0].replace("% overall alignment rate\n",""))
    rates.append(rate)

# add values to empty df for plotting
df["sample"] = samples
df["alignment_rate"] = rates

# plot alignment rate
utils.plot(df,"Overall alignment rate (%)", snakemake.output[0])

File: workflow/scripts/plot_alignment_rate.R
# redirect R output to log
log <- file(snakemake@log[[1]], open = "wt")
sink(log, type = "output")
sink(log, type = "message")

library(tidyverse)
library(cowplot)

# get HISAT2 alignment rates (in log files)
files <- snakemake@input

# create df for storing alignment rates
df <- as.data.frame(matrix(ncol = 2, nrow = 0))
names(df) <- c("sample","alignment.rate")

# get sample name and mapping rates from log files
for (i in seq(files)){
  sample <- system(paste0("echo ", files[i], "| sed 's/.log//'"), intern=TRUE)
  sample <- basename(sample)
  
  rate <- system(paste0('grep "aligned exactly 1 time" ', files[i], " | awk '{print $2}' | sed 's/(*[%)]*//g'"), intern=TRUE)
  rate <- as.numeric(rate)

  # add to df
  df[i,"sample"] <- sample
  df[i,"mapping.rate"] <- rate
}

# remove prepended X from samples names (only happens when they start with a number)
df$sample <- str_remove(df$sample, "^X")

# create plot
p <- ggplot(df, aes(x = sample, y = mapping.rate)) +
  geom_bar(stat = "identity", 
           fill = "aquamarine4",
           colour = "black") +
  theme_cowplot(16) +
  scale_y_continuous(limits = c(0,100)) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  ylab("Mapping rate (%)") +
  xlab(NULL)

# save plot
ggsave(snakemake@output[[1]], p)

# close log file
sink(log, type = "output")
sink(log, type = "message")
